File: app/src/main/java/com/example/weatherapp/MainActivity.java
package com.example.weatherapp;

import android.Manifest;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.KeyEvent;
import android.view.inputmethod.EditorInfo;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import java.util.List;

public class MainActivity extends AppCompatActivity {

    private static final int REQ_LOCATION = 101;

    // View
    private TextView textLocation, txtTemp, txtDesc, txtHumidity, txtWind, txtPressure;
    private EditText editCity;
    private Button btnSearch;
    private ImageView imgIcon;
    private RecyclerView recyclerForecast;

    // Adapter 7 hari (asumsikan sudah ada ForecastAdapter + ForecastDay)
    private ForecastAdapter forecastAdapter;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        bindViews();
        setupRecycler();
        setupSearch();

        // Coba load kota terakhir; jika tidak ada -> auto lokasi
        loadLastOrDetectLocation();
    }

    private void bindViews() {
        textLocation   = findViewById(R.id.textLocation);
        txtTemp        = findViewById(R.id.txtTemp);
        txtDesc        = findViewById(R.id.txtDesc);
        txtHumidity    = findViewById(R.id.txtHumidity);
        txtWind        = findViewById(R.id.txtWind);
        txtPressure    = findViewById(R.id.txtPressure);
        imgIcon        = findViewById(R.id.imgIcon);

        editCity       = findViewById(R.id.editCity);
        btnSearch      = findViewById(R.id.btnSearch);
        recyclerForecast = findViewById(R.id.recyclerForecast);
    }

    private void setupRecycler() {
        // Grid 7 kolom (7-day)
        recyclerForecast.setLayoutManager(new GridLayoutManager(this, 7));
        forecastAdapter = new ForecastAdapter(new java.util.ArrayList<>());
        recyclerForecast.setAdapter(forecastAdapter);
    }

    private void setupSearch() {
        btnSearch.setOnClickListener(v -> trySearchCity());
        editCity.setOnEditorActionListener((v, actionId, event) -> {
            boolean ime = actionId == EditorInfo.IME_ACTION_SEARCH;
            boolean enter = event != null && event.getKeyCode() == KeyEvent.KEYCODE_ENTER
                    && event.getAction() == KeyEvent.ACTION_DOWN;
            if (ime || enter) {
                trySearchCity();
                return true;
            }
            return false;
        });
    }

    private void trySearchCity() {
        final String q = editCity.getText() != null ? editCity.getText().toString().trim() : "";
        if (TextUtils.isEmpty(q)) {
            Toast.makeText(this, "Masukkan nama kota", Toast.LENGTH_SHORT).show();
            return;
        }
        // Cari lat/lon via Open-Meteo Geocoding
        WeatherServiceOM.searchCityLatLon(q, new WeatherServiceOM.Result<GeoResponse.Result>() {
            @Override public void ok(GeoResponse.Result res) {
                // Simpan ke Room (opsional — kamu sudah punya City/CityDao/WeatherDatabase)
                new Thread(() -> {
                    try {
                        WeatherDatabase.getInstance(MainActivity.this)
                                .cityDao().save(new City(res.name, res.latitude, res.longitude));
                    } catch (Throwable ignored) {}
                }).start();

                // Load cuaca 7 hari
                loadFromLatLon(res.latitude, res.longitude, res.name);
            }
            @Override public void err(Throwable t) {
                Toast.makeText(MainActivity.this, "Kota tidak ditemukan", Toast.LENGTH_SHORT).show();
            }
        });
    }

    private void loadLastOrDetectLocation() {
        // Ambil kota terakhir dari Room
        new Thread(() -> {
            City last = null;
            try {
                last = WeatherDatabase.getInstance(MainActivity.this).cityDao().getLastCity();
            } catch (Throwable ignored) {}

            final City lastCity = last;
            runOnUiThread(() -> {
                if (lastCity != null) {
                    loadFromLatLon(lastCity.lat, lastCity.lon, lastCity.name);
                } else {
                    requestOrFetchLocation();
                }
            });
        }).start();
    }

    private void requestOrFetchLocation() {
        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED &&
                ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this,
                    new String[]{Manifest.permission.ACCESS_FINE_LOCATION, Manifest.permission.ACCESS_COARSE_LOCATION},
                    REQ_LOCATION);
            return;
        }
        // Izin sudah OK -> ambil lokasi terakhir
        new LocationHelper(this).lastKnown(new LocationHelper.Callback() {
            @Override public void onLocation(double lat, double lon) {
                loadFromLatLon(lat, lon, "Lokasi Saya");
            }
            @Override public void onError(Throwable t) {
                Toast.makeText(MainActivity.this, "Gagal ambil lokasi", Toast.LENGTH_SHORT).show();
            }
        });
    }

    // Ambil current + 7 hari dari Open-Meteo dan render ke UI
    private void loadFromLatLon(double lat, double lon, String displayName) {
        WeatherServiceOM.fetch7Days(lat, lon, new WeatherServiceOM.Result<WeatherResponseOM>() {
            @Override public void ok(WeatherResponseOM data) {
                renderCurrent(displayName, data);
                render7Days(data);
            }
            @Override public void err(Throwable t) {
                Toast.makeText(MainActivity.this, "Gagal memuat cuaca", Toast.LENGTH_SHORT).show();
            }
        });
    }

    private void renderCurrent(String name, WeatherResponseOM data) {
        textLocation.setText(name);

        if (data != null && data.current != null) {
            long t = Math.round(data.current.temperature_2m);
            txtTemp.setText(t + "°C");
            txtDesc.setText(WeatherIconMapper.describe(data.current.weather_code));

            int hum = data.current.relative_humidity_2m != null
                    ? data.current.relative_humidity_2m.intValue()
                    : 0;

            int wind = data.current.wind_speed_10m != null
                    ? (int) Math.round(data.current.wind_speed_10m)
                    : 0;

            int pres = data.current.pressure_msl != null
                    ? (int) Math.round(data.current.pressure_msl)
                    : 0;

            txtHumidity.setText(hum + "%");
            txtWind.setText(wind + " km/h");
            txtPressure.setText(pres + " hPa");

            // (opsional) ganti ikon berdasarkan weather_code jika kamu punya drawable mapping sendiri
            // imgIcon.setImageResource( ... );
        }
    }

    private void render7Days(WeatherResponseOM data) {
        if (data == null || data.daily == null || data.daily.time == null) {
            forecastAdapter.submit(java.util.Collections.emptyList());
            return;
        }
        List<ForecastDay> items = WeatherMapper.toForecastDays(data.daily);
        forecastAdapter.submit(items);
    }

    // Callback permission lokasi
    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] perms, @NonNull int[] res) {
        super.onRequestPermissionsResult(requestCode, perms, res);
        if (requestCode == REQ_LOCATION) {
            if (res.length > 0 && res[0] == PackageManager.PERMISSION_GRANTED) {
                requestOrFetchLocation();
            } else {
                Toast.makeText(this, "Izin lokasi ditolak", Toast.LENGTH_SHORT).show();
            }
        }
    }
}

File: app/src/main/java/com/example/weatherapp/MainActivity.kt
package com.example.weatherapp

import android.Manifest
import android.annotation.SuppressLint
import android.content.ActivityNotFoundException
import android.content.Intent
import android.content.SharedPreferences
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.provider.Settings
import android.widget.ArrayAdapter
import android.widget.Toast
import androidx.activity.ComponentActivity
import androidx.activity.result.contract.ActivityResultContracts
import androidx.activity.viewModels
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatDelegate
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import androidx.core.content.ContextCompat
import androidx.core.widget.addTextChangedListener
import androidx.recyclerview.widget.GridLayoutManager
import com.example.weatherapp.databinding.ActivityMainBinding
import com.google.android.material.chip.Chip
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.cancel
import kotlinx.coroutines.flow.collectLatest
import kotlinx.coroutines.launch
import kotlin.math.roundToInt

class MainActivity : ComponentActivity() {

    // ViewModel, repo & adapter cuaca
    private val vm: MainViewModel by viewModels()
    private lateinit var binding: ActivityMainBinding
    private val repo = WeatherRepository()
    private val gridAdapter = DailyAdapter()

    // Scope untuk collect Flow
    private val scope = MainScope()

    // State suggestion & notifikasi
    private var lastSuggestions: List<GeoResult> = emptyList()
    private var lastNotifiedQuery: String = ""
    private var pendingAutoNotify: Boolean = false

    // ------------------------------
    //        RIWAYAT PENCARIAN
    // ------------------------------
    private lateinit var prefs: SharedPreferences
    private val searchHistory: MutableList<String> = mutableListOf()

    private val PREF_NAME = "weather_prefs"
    private val KEY_HISTORY = "search_history"
    private val MAX_HISTORY_SIZE = 10
    // ------------------------------

    // Launcher izin notifikasi
    private val requestNotifPermission = registerForActivityResult(
        ActivityResultContracts.RequestPermission()
    ) { granted ->
        if (granted && pendingAutoNotify) {
            showWeatherNotificationSafely()
            pendingAutoNotify = false
        } else if (!granted) {
            if (shouldShowPostNotifRationale()) showPostNotifRationale()
            else showGoToSettingsDialog()
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)

        // Inisialisasi SharedPreferences & load riwayat
        prefs = getSharedPreferences(PREF_NAME, MODE_PRIVATE)
        loadSearchHistoryFromPrefs()
        renderSearchHistoryChips()

        // RecyclerView 7-day forecast
        binding.recyclerForecast.apply {
            layoutManager = GridLayoutManager(this@MainActivity, 7)
            adapter = gridAdapter
        }

        // Dropdown autocomplete
        val dropdownAdapter =
            ArrayAdapter<String>(this, android.R.layout.simple_list_item_1, mutableListOf())
        binding.editCity.setAdapter(dropdownAdapter)

        // Hint "Search Region" hilang saat user mengetik
        binding.editCity.addTextChangedListener { text ->
            binding.tilCity.hint = if (text.isNullOrEmpty()) "Search Region" else ""
            vm.onQueryChanged(text?.toString().orEmpty())
        }

        // User memilih suggestion dari dropdown
        binding.editCity.setOnItemClickListener { _, _, pos, _ ->
            val geo = lastSuggestions.getOrNull(pos) ?: return@setOnItemClickListener
            vm.selectLocation(geo)

            val selectedText = "${geo.name}, ${geo.country}"
            binding.editCity.setText(selectedText)
            binding.editCity.setSelection(selectedText.length)

            addQueryToHistory(selectedText)

            scope.launch {
                vm.forecast.collectLatest { f ->
                    if (f != null) ensureNotificationPermissionThenNotify()
                }
            }
        }

        // Tombol "Cari" ditekan
        binding.btnSearch.setOnClickListener {
            val q = binding.editCity.text?.toString().orEmpty()
            vm.onQueryChanged(q)
            addQueryToHistory(q)
        }

        // Observasi suggestions (kota)
        scope.launch {
            vm.suggestions.collectLatest { list ->
                lastSuggestions = list
                dropdownAdapter.clear()
                dropdownAdapter.addAll(list.map { "${it.name}, ${it.country}" })
                dropdownAdapter.notifyDataSetChanged()

                if (list.isNotEmpty()) binding.editCity.showDropDown()

                val q = binding.editCity.text?.toString().orEmpty()
                if (q.length >= 3 && list.isNotEmpty() && q != lastNotifiedQuery) {
                    lastNotifiedQuery = q
                    pendingAutoNotify = true
                    vm.selectLocation(list.first())

                    // opsional: simpan suggestion pertama sebagai riwayat
                    val firstText = "${list.first().name}, ${list.first().country}"
                    addQueryToHistory(firstText)
                }
            }
        }

        // Observasi forecast cuaca
        scope.launch {
            vm.forecast.collectLatest { f ->
                val night = repo.isNight(f?.current_weather?.is_day)
                AppCompatDelegate.setDefaultNightMode(
                    if (night) AppCompatDelegate.MODE_NIGHT_YES
                    else AppCompatDelegate.MODE_NIGHT_NO
                )

                val cur = f?.current_weather
                val icon = repo.iconFor(cur?.weathercode ?: 3, night)
                binding.imgIcon.setImageResource(icon)

                // Animasi icon
                binding.imgIcon.animate()
                    .rotationBy(360f)
                    .setDuration(600L)
                    .start()

                binding.txtTemp.text =
                    cur?.temperature?.roundToInt()?.let { "$it°C" } ?: "--°C"
                binding.textLocation.text =
                    binding.editCity.text?.toString().orEmpty().ifEmpty { "--" }

                binding.txtDesc.text = when (cur?.weathercode) {
                    0 -> if (night) "Cerah (malam)" else "Cerah"
                    in 1..3 -> if (night) "Cerah berawan (malam)" else "Cerah berawan"
                    in 61..65, in 51..67, in 80..82 -> "Hujan"
                    else -> "Berawan"
                }

                // Background dinamis
                if (cur != null) {
                    updateBackgroundWithAnimation(
                        isNight = night,
                        weatherCode = cur.weathercode
                    )
                }

                val hourly = f?.hourly
                val hum = cur?.time?.let {
                    repo.matchHourlyValue(
                        it,
                        hourly?.time,
                        hourly?.relativehumidity_2m
                    )
                }
                val pres = cur?.time?.let {
                    repo.matchHourlyValue(
                        it,
                        hourly?.time,
                        hourly?.surface_pressure
                    )
                }
                binding.txtHumidity.text = hum?.roundToInt()?.let { "$it%" } ?: "--%"
                binding.txtPressure.text = pres?.roundToInt()?.let { "$it hPa" } ?: "-- hPa"
                binding.txtWind.text =
                    cur?.windspeed?.roundToInt()?.let { "$it km/h" } ?: "-- km/h"

                val d = f?.daily
                if (d != null && d.time.isNotEmpty()) {
                    val items = d.time.indices.map { i ->
                        GridDay(
                            dateIso = d.time[i],
                            iconRes = repo.iconFor(d.weathercode[i], night),
                            tMax = d.temperature_2m_max[i],
                            tMin = d.temperature_2m_min[i]
                        )
                    }.take(7)
                    gridAdapter.submit(items)
                } else gridAdapter.submit(emptyList())

                if (pendingAutoNotify && f != null) {
                    ensureNotificationPermissionThenNotify()
                    pendingAutoNotify = false
                }
            }
        }
    }

    // --------------------------------------------------------------------
    //                    RIWAYAT PENCARIAN
    // --------------------------------------------------------------------

    /** Load riwayat dari SharedPreferences ke list */
    private fun loadSearchHistoryFromPrefs() {
        val raw = prefs.getString(KEY_HISTORY, "") ?: ""
        if (raw.isNotEmpty()) {
            val items = raw.split("|").filter { it.isNotBlank() }
            searchHistory.clear()
            searchHistory.addAll(items)
        }
    }

    /** Simpan riwayat ke SharedPreferences */
    private fun saveSearchHistoryToPrefs() {
        val joined = searchHistory.joinToString("|")
        prefs.edit().putString(KEY_HISTORY, joined).apply()
    }

    /** Tambahkan query ke riwayat + batasi ukuran list */
    private fun addQueryToHistory(query: String) {
        val q = query.trim()
        if (q.isEmpty()) return

        // hapus jika sudah ada (biar tidak dobel)
        searchHistory.remove(q)
        // tambahkan di paling depan (terbaru)
        searchHistory.add(0, q)

        // batasi jumlah maksimum riwayat (pakai removeAt, aman untuk API 24+)
        while (searchHistory.size > MAX_HISTORY_SIZE) {
            searchHistory.removeAt(searchHistory.size - 1)
        }

        saveSearchHistoryToPrefs()
        renderSearchHistoryChips()
    }

    /** Render searchHistory jadi chip di chipGroupHistory */
    private fun renderSearchHistoryChips() {
        val group = binding.chipGroupHistory
        group.removeAllViews()

        searchHistory.forEach { text ->
            val chip = Chip(this).apply {
                this.text = text
                isCheckable = false
                isClickable = true

                setOnClickListener {
                    binding.editCity.setText(text)
                    binding.editCity.setSelection(text.length)
                    vm.onQueryChanged(text)
                }
            }
            group.addView(chip)
        }
    }

    // --------------------------------------------------------------------
    //           BACKGROUND + IZIN & NOTIFIKASI CUACA
    // --------------------------------------------------------------------

    /** Background berubah (siang / malam / hujan) + animasi */
    private fun updateBackgroundWithAnimation(isNight: Boolean, weatherCode: Int) {
        val isRain = when (weatherCode) {
            in 51..67,
            in 80..82,
            in 61..65 -> true
            else -> false
        }

        val bgRes = when {
            isRain -> R.drawable.bg_weather_rain
            isNight -> R.drawable.bg_weather_gradient_night
            else -> R.drawable.bg_weather_gradient
        }

        val root = binding.rootContainer
        root.animate().cancel()
        root.alpha = 0f
        root.setBackgroundResource(bgRes)
        root.animate()
            .alpha(1f)
            .setDuration(600L)
            .start()
    }

    private fun needsPostNotifPermission(): Boolean =
        Build.VERSION.SDK_INT >= 33 &&
                ContextCompat.checkSelfPermission(
                    this,
                    Manifest.permission.POST_NOTIFICATIONS
                ) != PackageManager.PERMISSION_GRANTED

    private fun shouldShowPostNotifRationale(): Boolean =
        Build.VERSION.SDK_INT >= 33 &&
                shouldShowRequestPermissionRationale(Manifest.permission.POST_NOTIFICATIONS)

    private fun showPostNotifRationale() {
        AlertDialog.Builder(this)
            .setTitle("Izin Notifikasi")
            .setMessage("Aplikasi butuh izin untuk menampilkan notifikasi cuaca terbaru.")
            .setPositiveButton("Izinkan") { d, _ ->
                d.dismiss()
                requestNotifPermission.launch(Manifest.permission.POST_NOTIFICATIONS)
            }
            .setNegativeButton("Nanti") { d, _ -> d.dismiss() }
            .show()
    }

    private fun showGoToSettingsDialog() {
        AlertDialog.Builder(this)
            .setTitle("Aktifkan Notifikasi")
            .setMessage("Notifikasi dimatikan. Aktifkan melalui pengaturan aplikasi.")
            .setPositiveButton("Buka Pengaturan") { d, _ ->
                d.dismiss()
                openAppNotificationSettings()
            }
            .setNegativeButton("Batal") { d, _ -> d.dismiss() }
            .show()
    }

    private fun openAppNotificationSettings() {
        try {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                startActivity(Intent(Settings.ACTION_APP_NOTIFICATION_SETTINGS).apply {
                    putExtra(Settings.EXTRA_APP_PACKAGE, packageName)
                })
            } else {
                startActivity(Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS).apply {
                    data = Uri.fromParts("package", packageName, null)
                })
            }
        } catch (_: ActivityNotFoundException) {
            startActivity(Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS).apply {
                data = Uri.fromParts("package", packageName, null)
            })
        }
    }

    private fun ensureNotificationPermissionThenNotify() {
        if (Build.VERSION.SDK_INT < 33) {
            showWeatherNotificationSafely()
            return
        }

        if (needsPostNotifPermission()) {
            pendingAutoNotify = true
            if (shouldShowPostNotifRationale()) showPostNotifRationale()
            else requestNotifPermission.launch(Manifest.permission.POST_NOTIFICATIONS)
            return
        }

        if (!NotificationManagerCompat.from(this).areNotificationsEnabled()) {
            showGoToSettingsDialog()
            return
        }

        showWeatherNotificationSafely()
    }

    private fun showWeatherNotificationSafely() {
        try {
            showWeatherNotification()
        } catch (_: SecurityException) {
            Toast.makeText(this, "Izin notifikasi ditolak.", Toast.LENGTH_SHORT).show()
        }
    }

    // Notifikasi cuaca
    @SuppressLint("MissingPermission")
    private fun showWeatherNotification() {
        val f = vm.forecast.value ?: return
        val cur = f.current_weather ?: return
        val night = cur.is_day == 0
        val icon = repo.iconFor(cur.weathercode, night)
        val text =
            "Suhu ${cur.temperature.roundToInt()}°C • Angin ${cur.windspeed.roundToInt()} km/j • Kode ${cur.weathercode}"

        val notif = NotificationCompat.Builder(this, "weather_detail")
            .setSmallIcon(icon)
            .setContentTitle("Cuaca saat ini")
            .setContentText(text)
            .setStyle(NotificationCompat.BigTextStyle().bigText(text))
            .setAutoCancel(true)
            .build()

        NotificationManagerCompat.from(this).notify(1001, notif)
    }

    override fun onDestroy() {
        super.onDestroy()
        scope.cancel()
    }
}
